internal/api/oidc/userinfo.go

```diff
@@ -0,0 +1,276 @@
+package oidc
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"slices"
+	"strings"
+
+	"github.com/dop251/goja"
+	"github.com/zitadel/logging"
+	"github.com/zitadel/oidc/v3/pkg/oidc"
+
+	"github.com/zitadel/zitadel/internal/actions"
+	"github.com/zitadel/zitadel/internal/actions/object"
+	"github.com/zitadel/zitadel/internal/domain"
+	"github.com/zitadel/zitadel/internal/query"
+)
+
+func (s *Server) userInfo(ctx context.Context, userID, projectID string, scope, roleAudience []string) (_ *oidc.UserInfo, err error) {
+	roleAudience, requestedRoles := prepareRoles(ctx, projectID, scope, roleAudience)
+	qu, err := s.query.GetOIDCUserInfo(ctx, userID, roleAudience)
+	if err != nil {
+		return nil, err
+	}
+
+	userInfo := userInfoToOIDC(projectID, qu, scope, roleAudience, requestedRoles, s.assetAPIPrefix(ctx))
+	return userInfo, s.userinfoFlows(ctx, qu, userInfo)
+}
+
+// prepareRoles scans the requested scopes, appends to roleAudiendce and returns the requestedRoles.
+//
+// When [ScopeProjectsRoles] is present and roleAudience was empty,
+// project IDs with the [domain.ProjectIDScope] prefix are added to the roleAudience.
+//
+// Scopes with [ScopeProjectRolePrefix] are added to requestedRoles.
+//
+// If the resulting requestedRoles or roleAudience are not not empty,
+// the current projectID will always be parts or roleAudience.
+// Else nil, nil is returned.
+func prepareRoles(ctx context.Context, projectID string, scope, roleAudience []string) (ra, requestedRoles []string) {
+	// if all roles are requested take the audience for those from the scopes
+	if slices.Contains(scope, ScopeProjectsRoles) && len(roleAudience) == 0 {
+		roleAudience = domain.AddAudScopeToAudience(ctx, roleAudience, scope)
+	}
+	requestedRoles = make([]string, 0, len(scope))
+	for _, s := range scope {
+		if role, ok := strings.CutPrefix(s, ScopeProjectRolePrefix); ok {
+			requestedRoles = append(requestedRoles, role)
+		}
+	}
+	if len(requestedRoles) == 0 && len(roleAudience) == 0 {
+		return nil, nil
+	}
+
+	if projectID != "" && !slices.Contains(roleAudience, projectID) {
+		roleAudience = append(roleAudience, projectID)
+	}
+	return roleAudience, requestedRoles
+}
+
+func userInfoToOIDC(projectID string, user *query.OIDCUserInfo, scope, roleAudience, requestedRoles []string, assetPrefix string) *oidc.UserInfo {
+	out := new(oidc.UserInfo)
+	for _, s := range scope {
+		switch s {
+		case oidc.ScopeOpenID:
+			out.Subject = user.User.ID
+		case oidc.ScopeEmail:
+			out.UserInfoEmail = userInfoEmailToOIDC(user.User)
+		case oidc.ScopeProfile:
+			out.UserInfoProfile = userInfoProfileToOidc(user.User, assetPrefix)
+		case oidc.ScopePhone:
+			out.UserInfoPhone = userInfoPhoneToOIDC(user.User)
+		case oidc.ScopeAddress:
```

Comment:
Should we remove this as not used?

Reply:
The switch case was duplicated from here:

https://github.com/zitadel/zitadel/blob/2e2214339ed3925dbdc52e44eeea631ab3562f32/internal/api/oidc/client.go#L350-L351

That particular code will be removed once we port the userinfo endpoint to the new code and the TODO will be lost if not kept here.

I've consulted @livio-a around the matter last week. I understood that there was a disccusion around the implementation of the address scope between the zitadel engineers way before my time that never got resolved. I also cannot find any issue or user story on the follow up.

I can remove the TODO, but then the engineers involved should at least define an userstory with the needed functionality.

Reply:
Ok, jsut wanted to point it out, as it was included in the PR and I was not sure what to do with it.